### PR TITLE
20197: Containerized Android builds

### PR DESCRIPTION
**Pull request by jherico**
[20197](https://highfidelity.manuscript.com/f/cases/20197/Migrate-Android-builds-to-containers)

The primary task of this PR is to make the CI build system for Android more configurable by the Android developers without having to get the dev ops team involved.  By moving to a containerized build, Android developers will have direct control over the environment and installed tools that will be used to build Android apps, not only on their own systems, but on the CI build servers as well.  

* Containerize the Android build by using a Dockerfile
* Add the Gradle wrapper script to the Android project
* Update the NDK version to r18

This is required for future work supporting Quest since we want to update the gradle version and NDK version for the Android build system prior to the Quest build.  

## Testing

This slightly modifies the pre-build script and the android build process. The android build should be smoke tested as this updates the Android NDK to r18.  The non-Android targets should be fine as long as they continue to successfully build.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15588/HighFidelity-Beta-PR14521-30efe58.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15588/HighFidelity-Beta-PR14521-30efe58.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15588/HighFidelity-Beta-Interface-PR14521-30efe58.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15588/HighFidelity-Beta-PR14521-30efe58.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15588/HighFidelity-Beta-Interface-PR14521-30efe58.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15593/HighFidelity-Beta-PR14521-39fb989.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15593/HighFidelity-Beta-PR14521-39fb989.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15593/HighFidelity-Beta-Interface-PR14521-39fb989.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15604/HighFidelity-Beta-PR14521-58bb9a2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15604/HighFidelity-Beta-Interface-PR14521-58bb9a2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15604/HighFidelity-Beta-PR14521-58bb9a2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15604/HighFidelity-Beta-Interface-PR14521-58bb9a2.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15606/HighFidelity-Beta-PR14521-6ff01db.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15606/HighFidelity-Beta-Interface-PR14521-6ff01db.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15609/HighFidelity-Beta-PR14521-8281b2d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15609/HighFidelity-Beta-Interface-PR14521-8281b2d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15609/HighFidelity-Beta-PR14521-8281b2d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15609/HighFidelity-Beta-Interface-PR14521-8281b2d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15614/HighFidelity-Beta-PR14521-8639537.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15614/HighFidelity-Beta-Interface-PR14521-8639537.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15621/HighFidelity-Beta-PR14521-4b84373.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15621/HighFidelity-Beta-Interface-PR14521-4b84373.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15621/HighFidelity-Beta-PR14521-4b84373.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15621/HighFidelity-Beta-Interface-PR14521-4b84373.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15628/HighFidelity-Beta-PR14521-c4d53e0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15628/HighFidelity-Beta-Interface-PR14521-c4d53e0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15628/HighFidelity-Beta-PR14521-c4d53e0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15628/HighFidelity-Beta-Interface-PR14521-c4d53e0.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15643/HighFidelity-Beta-PR14521-be9e777.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15643/HighFidelity-Beta-Interface-PR14521-be9e777.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15649/HighFidelity-Beta-PR14521-4d9eb2e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15649/HighFidelity-Beta-Interface-PR14521-4d9eb2e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15653/HighFidelity-Beta-PR14521-253f86f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15653/HighFidelity-Beta-Interface-PR14521-253f86f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15653/HighFidelity-Beta-PR14521-253f86f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15653/HighFidelity-Beta-Interface-PR14521-253f86f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15658/HighFidelity-Beta-PR14521-fb415de.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15658/HighFidelity-Beta-PR14521-fb415de.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15658/HighFidelity-Beta-Interface-PR14521-fb415de.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15658/HighFidelity-Beta-PR14521-fb415de.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15658/HighFidelity-Beta-Interface-PR14521-fb415de.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15666/HighFidelity-Beta-PR14521-8d6b575.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15666/HighFidelity-Beta-PR14521-8d6b575.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15666/HighFidelity-Beta-Interface-PR14521-8d6b575.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15666/HighFidelity-Beta-PR14521-8d6b575.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15666/HighFidelity-Beta-Interface-PR14521-8d6b575.exe

File: prebuild.py [hunk 1]
```diff
@@ -9,252 +30,65 @@
...
-parser.add_argument('--build-root', required=True, type=str, help='The location of the cmake build')
-
-args = parser.parse_args()
+    # Fixup env variables.  Leaving `USE_CCACHE` on will cause scribe to fail to build
+    # VCPKG_ROOT seems to cause confusion on Windows systems that previously used it for 
+    # building OpenSSL
+    removeEnvVars = ['VCPKG_ROOT', 'USE_CCACHE']
+    for var in removeEnvVars:
+        if var in os.environ:
+            del os.environ[var]
```
Comment (Atlante45) [inline prebuild.py]: Does this propagate to the cmake command running afterward?
If so, we should restore them at the end.

Comment (jherico) [reply prebuild.py]: cmake calls the python script, not the other way around.  so this doesn't affect the cmake stuff

**Comment by qligmaloney** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15833/HighFidelity-Beta-PR14521-8d6b575.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15833/HighFidelity-Beta-Interface-PR14521-8d6b575.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15840/HighFidelity-Beta-PR14521-39cff4f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15840/HighFidelity-Beta-PR14521-39cff4f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15840/HighFidelity-Beta-Interface-PR14521-39cff4f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15840/HighFidelity-Beta-PR14521-39cff4f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15840/HighFidelity-Beta-Interface-PR14521-39cff4f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15849/HighFidelity-Beta-PR14521-5cc00be.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15849/HighFidelity-Beta-PR14521-5cc00be.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15849/HighFidelity-Beta-Interface-PR14521-5cc00be.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15849/HighFidelity-Beta-PR14521-5cc00be.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15849/HighFidelity-Beta-Interface-PR14521-5cc00be.exe

**Comment by marko8904** [comment]
@jherico This is awesome! Happy to see the containerized builds for Android! We've discussed the changes you've made and I'm on-board. 👍 

Just a suggestion. I think that a readme next to Dockerfile would be helpful to engineers new to Docker who wish to build for android locally (in the same way that the build pipeline does).

**Comment by qaAdamSmith82** [comment]
Merged without QA.  Will run the test plan against master.

**Pull request merged**